applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,354 @@
+
+	return 0;
+}
+
+static int disable_interrupts(struct selector *selector)
+{
+	const struct pin *sel_pins = pins[selector->id];
+	int key = irq_lock();
+
+	for (size_t i = 0; i < selector->pin_cnt; i++) {
+		int err = gpio_pin_disable_callback(gpio_dev[sel_pins[i].port],
+						    sel_pins[i].pin);
+
+		if (err) {
+			LOG_ERR("Cannot disable callback (err %d)", err);
+			return err;
+		}
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+void selector_work_fn(struct k_work *w)
+{
+	struct selector *selector = CONTAINER_OF(w, struct selector, work);
+
+	selector_event_send(selector);
+
+	int err = enable_interrupts(selector);
+
+	if (err) {
+		module_set_state(MODULE_STATE_ERROR);
+	}
+}
+
+static int configure(struct selector *selector)
+{
+	const struct pin *sel_pins = pins[selector->id];
+	u32_t bitmask = 0;
+
+	for (size_t i = 0; i < ARRAY_SIZE(pins[selector->id]); i++) {
+		if ((sel_pins[i].port == 0) && (sel_pins[i].pin == 0)) {
+			__ASSERT_NO_MSG(i != 0);
+
+			selector->pin_cnt = i;
+			break;
+		}
+	}
+	if (selector->pin_cnt == 0) {
+		selector->pin_cnt = ARRAY_SIZE(pins[selector->id]);
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(gpio_dev); i++) {
+		for (size_t j = 0; j < selector->pin_cnt; j++) {
+			if (sel_pins[j].port == i) {
+				bitmask |= BIT(sel_pins[j].pin);
+			}
+		}
+		gpio_init_callback(&selector->gpio_cb[i], selector_isr,
+				   bitmask);
+		int err = gpio_add_callback(gpio_dev[i], &selector->gpio_cb[i]);
+
+		if (err) {
+			LOG_ERR("Cannot add callback (err %d)", err);
+			return err;
+		}
+
+		bitmask = 0;
+	}
+
+	for (size_t i = 0; i < selector->pin_cnt; i++) {
+		int err = gpio_pin_configure(gpio_dev[sel_pins[i].port],
+					     sel_pins[i].pin,
+					     GPIO_PUD_PULL_DOWN | GPIO_DIR_IN |
+					     GPIO_INT | GPIO_INT_LEVEL |
+					     GPIO_INT_ACTIVE_HIGH);
+
+		if (err) {
+			LOG_ERR("Cannot configure pin (err %d)", err);
+			return err;
+		}
+	}
+
+	k_delayed_work_init(&selector->work, selector_work_fn);
+
+	return 0;
+}
+
+static int init(void)
+{
+	int err;
+
+	for (size_t i = 0; i < ARRAY_SIZE(port_map); i++) {
+		gpio_dev[i] = device_get_binding(port_map[i]);
+
+		if (!gpio_dev[i]) {
+			LOG_ERR("Cannot get GPIO%u dev binding", i);
+			return -1;
+		}
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(selectors); i++) {
+		selectors[i].id = i;
+		err = configure(&selectors[i]);
+		if (err) {
+			return err;
+		}
+
+		err = read_state(&selectors[i]);
+		if (err) {
+			return err;
+		}
+
+		err = enable_interrupts(&selectors[i]);
+		if (err) {
+			return err;
+		}
+	}
+	state = STATE_ACTIVE;
+
+	return 0;
+}
+
+static void sleep(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(selectors); i++) {
+		int err = disable_interrupts(&selectors[i]);
+
+		if (err) {
+			module_set_state(MODULE_STATE_ERROR);
+			return;
+		}
+
+		k_delayed_work_cancel(&selectors[i].work);
+	}
+
+	state = STATE_OFF;
+	module_set_state(MODULE_STATE_OFF);
+}
+
+static void wake_up(void)
+{
+	for (size_t i = 0; ARRAY_SIZE(selectors); i++) {
```

Comment:
similar code is part of init function, the difference is that, the init does pins configuration.

Reply:
I could split it to configuration and enabling pins (2 for loops in 2 separate functions). Will it be ok ?

Reply:
I wanted to suggest to extract code that is duplicated into a separate function.

Reply:
Ok, I will add function for this